### ISSUE TITLE
Running multiple pipelines in Kedro

### ISSUE THREAD

**Issue by sanjaymeena** [quantumblacklabs/kedro#396]
Hello, I am using Kedro for a project which I want to extend to support different countries. I want to have one pipeline for each country in the project.  Any pointers on how can I achieve this ?

**Comment by mzjp2** [comment]
Hey @sanjaymeena!

I wonder if this is a usecase for some of the [modular pipeline](https://kedro.readthedocs.io/en/stable/04_user_guide/06_pipelines.html#developing-modular-pipelines) functionality we have in the latest Kedro release. @lorenabalan has an answer here to a separate question that I think could be useful in your case to look at as a reference to implementing your multiple pipelines: https://github.com/quantumblacklabs/kedro/issues/162#issuecomment-635257763

The documentation for these can be found here: https://kedro.readthedocs.io/en/stable/04_user_guide/06_pipelines.html#developing-modular-pipelines

**Comment by sanjaymeena** [comment]
Hey @mzjp2  , thanks! I've upgraded kedro version to `0.16.1` and trying out modular pipelines. They are nice! 

I had another question regarding common parameters across pipelines.  My two pipelines differ only by a country parameter which is defined in `conf/base/parameters.yaml`. I am unable to override that parameter in `conf/base/pipelines/parameters.yaml` as I get 

`ValueError: Duplicate keys found` error .

**Comment by lorenabalan** [comment]
Hey @sanjaymeena that's because the two parameters files reside in the same environment, `base`. The `ConfigLoader` will load up all `parameters` files and folders and collate them together into one, so it will detect duplicate keys.

Instead what you can do is refer to a different parameter in your pipeline. If you have your original pipeline using `params:UK_country_code`, using sth like this

```
pipeline(original_pipeline, inputs=..., outputs=..., parameters={"params:UK_country_code": "params:US_country_code", namespace="America")
```

and it will replace all references to UK_country_code parameter in your node inputs with the new parameter US_country_code. Hope this helps!

**Comment by sanjaymeena** [comment]
thanks @lorenabalan , for some reason I still got errors with this approach.  for now, I am passing country parameter at node level. 

Sorry to keep bugging with questions, I had another one ..

I previously used to pass `env` and `params` to Kedro run through `load_context` . For modular pipelines, I am using demo code which has `
`
project_context = load_package_context(
        project_path=Path.cwd(),
        package_name=Path(__file__).resolve().parent.name,
    )`

I am unable to pass on extra params from `kedro run --params COUNTRIES:['AU']`  . any pointers ?

**Comment by lorenabalan** [comment]
What's the error you are getting? Do you have both parameter names in your config file (the original one, and the one to replace it with)?

The code snippet you included is **not** what is run when you do `kedro run` from the CLI. It's only if you are running `run.py` as a script or you have packaged and pip installed the project and running it that way. `kedro run` executes `ProjectContext.run()` underneath. We encourage to pass env from the CLI as well `kedro run --env`.

As for parameters, the `--params` option indeed only accepts very basic types, so no lists like you are passing in your example. If you reaaaaally need it, you can tweak your `kedro_cli.py` to fit the preferred behaviour for this option, but we strongly encourage users to use config files and environments as much as possible. Extra params at run time are a special case.

In general [RELEASE.md](https://github.com/quantumblacklabs/kedro/blob/master/RELEASE.md) is the place to check. We include a break down of breaking changes and non-breaking changes, as well as migration guide for major version bumps: https://github.com/quantumblacklabs/kedro/blob/develop/RELEASE.md#migration-guide-from-kedro-015-to-016

**Comment by sanjaymeena** [comment]
thanks @lorenabalan, I started to use environment variables for setting params. My code gets triggered externally from gitlab ci/cd so previously I used to pass the --params and tweak `kedro_cli.py` .

**Comment by sanjaymeena** [comment]
Hello everyone, 

I had another question related to modular pipelines.  

I have defined my two country pipelines to run by concatenating them ( `pipeline_dict["__default__"] = `country1_pipeline` + `country2_pipeline`  `

I can run both the pipelines sequentially using `Kedro run ` without any problems.   However, it's not the case that all the nodes in `country1_pipeline` will finish processing before beginning processing of `country2_pipeline` . 

Is there a way I can make sure that `country1_pipeline` will finish processing before the execution of `country2_pipeline`  begins?

**Comment by lorenabalan** [comment]
Some sort of artificial linking nodes would do the trick. You can "force" the waiting by adding a node that takes `country1_pipeline`'s outputs as inputs, spits out some dummy output (e.g. a MemoryDataSet) that is pulled as input into `country2_pipeline`.

**Comment by lorenabalan** [comment]
Closing this as answered, but feel free to open a new issue or ask us something on StackOverflow (make sure you tag your questions with `kedro`).
